File: src/__tests__/crypto.test.ts
import { encryptValue, decryptValue, isEncrypted } from '../utils/crypto';

describe('Crypto utilities', () => {
  const password = 'test-password-123';

  describe('isEncrypted', () => {
    it('returns true for encrypted values', () => {
      const encrypted = encryptValue('hello', password);
      expect(isEncrypted(encrypted)).toBe(true);
    });

    it('returns false for plain values', () => {
      expect(isEncrypted('hello world')).toBe(false);
      expect(isEncrypted('')).toBe(false);
      expect(isEncrypted('encrypted::')).toBe(false);
      expect(isEncrypted('encrypted::v2:abc')).toBe(false);
    });

    it('returns true for the exact prefix format', () => {
      expect(isEncrypted('encrypted::v1:abc')).toBe(true);
    });
  });

  describe('encryptValue / decryptValue round-trip', () => {
    it('encrypts and decrypts a simple string', () => {
      const plaintext = 'sk-abc123';
      const encrypted = encryptValue(plaintext, password);
      expect(isEncrypted(encrypted)).toBe(true);
      expect(encrypted).not.toContain(plaintext);
      const decrypted = decryptValue(encrypted, password);
      expect(decrypted).toBe(plaintext);
    });

    it('handles unicode text', () => {
      const plaintext = 'Hello \u{1F30D} \u00E9\u00E8\u00EA \u4F60\u597D';
      const encrypted = encryptValue(plaintext, password);
      const decrypted = decryptValue(encrypted, password);
      expect(decrypted).toBe(plaintext);
    });

    it('handles empty string', () => {
      const encrypted = encryptValue('', password);
      const decrypted = decryptValue(encrypted, password);
      expect(decrypted).toBe('');
    });

    it('handles long values', () => {
      const plaintext = 'x'.repeat(10000);
      const encrypted = encryptValue(plaintext, password);
      const decrypted = decryptValue(encrypted, password);
      expect(decrypted).toBe(plaintext);
    });

    it('handles multi-line values', () => {
      const plaintext = 'line1\nline2\nline3';
      const encrypted = encryptValue(plaintext, password);
      const decrypted = decryptValue(encrypted, password);
      expect(decrypted).toBe(plaintext);
    });

    it('produces different ciphertexts for the same input (random salt/iv)', () => {
      const a = encryptValue('same', password);
      const b = encryptValue('same', password);
      expect(a).not.toBe(b);
      // Both should decrypt to the same value
      expect(decryptValue(a, password)).toBe('same');
      expect(decryptValue(b, password)).toBe('same');
    });
  });

  describe('decryptValue error handling', () => {
    it('throws on wrong password', () => {
      const encrypted = encryptValue('secret', password);
      expect(() => decryptValue(encrypted, 'wrong-password')).toThrow('Decryption failed');
    });

    it('throws on non-encrypted input', () => {
      expect(() => decryptValue('plain text', password)).toThrow('not encrypted');
    });

    it('throws on corrupted data (truncated payload)', () => {
      expect(() => decryptValue('encrypted::v1:abc', password)).toThrow('Corrupted');
    });

    it('throws on corrupted data (modified ciphertext)', () => {
      const encrypted = encryptValue('secret', password);
      // Modify a character in the base64 payload
      const prefix = 'encrypted::v1:';
      const payload = encrypted.slice(prefix.length);
      const corruptedPayload = payload.slice(0, -2) + 'XX';
      expect(() => decryptValue(prefix + corruptedPayload, password)).toThrow();
    });
  });
});

File: codexcli-go/internal/crypto/crypto_test.go
package crypto

import (
	"strings"
	"testing"
)

func TestIsEncrypted(t *testing.T) {
	tests := []struct {
		input string
		want  bool
	}{
		{"encrypted::v1:abc123", true},
		{"encrypted::v1:", true},
		{"plain text", false},
		{"encrypted::v2:abc", false},
		{"", false},
	}
	for _, tt := range tests {
		if got := IsEncrypted(tt.input); got != tt.want {
			t.Errorf("IsEncrypted(%q) = %v, want %v", tt.input, got, tt.want)
		}
	}
}

func TestEncryptDecryptRoundTrip(t *testing.T) {
	plaintext := "hello, world!"
	password := "test-password-123"

	encrypted, err := EncryptValue(plaintext, password)
	if err != nil {
		t.Fatalf("EncryptValue() error = %v", err)
	}

	if !IsEncrypted(encrypted) {
		t.Errorf("encrypted value should have prefix")
	}

	if !strings.HasPrefix(encrypted, "encrypted::v1:") {
		t.Errorf("expected prefix 'encrypted::v1:', got %q", encrypted[:20])
	}

	decrypted, err := DecryptValue(encrypted, password)
	if err != nil {
		t.Fatalf("DecryptValue() error = %v", err)
	}

	if decrypted != plaintext {
		t.Errorf("DecryptValue() = %q, want %q", decrypted, plaintext)
	}
}

func TestDecryptWrongPassword(t *testing.T) {
	encrypted, err := EncryptValue("secret", "correct-password")
	if err != nil {
		t.Fatalf("EncryptValue() error = %v", err)
	}

	_, err = DecryptValue(encrypted, "wrong-password")
	if err == nil {
		t.Error("expected decryption error with wrong password")
	}
}

func TestDecryptNotEncrypted(t *testing.T) {
	_, err := DecryptValue("plain text", "password")
	if err == nil {
		t.Error("expected error for non-encrypted value")
	}
}

func TestDecryptCorrupted(t *testing.T) {
	_, err := DecryptValue("encrypted::v1:invalidbase64!!!", "password")
	if err == nil {
		t.Error("expected error for corrupted data")
	}
}

func TestEncryptDifferentEachTime(t *testing.T) {
	plaintext := "same input"
	password := "same password"

	e1, _ := EncryptValue(plaintext, password)
	e2, _ := EncryptValue(plaintext, password)

	if e1 == e2 {
		t.Error("two encryptions of the same input should produce different output (random salt/iv)")
	}
}

func TestMaskEncryptedValues(t *testing.T) {
	input := map[string]any{
		"plain":  "visible",
		"secret": "encrypted::v1:abc123",
		"nested": map[string]any{
			"deep": "encrypted::v1:def456",
			"ok":   "fine",
		},
	}

	result := MaskEncryptedValues(input)

	if result["plain"] != "visible" {
		t.Errorf("plain value should be unchanged")
	}
	if result["secret"] != "[encrypted]" {
		t.Errorf("encrypted value should be masked")
	}
	nested := result["nested"].(map[string]any)
	if nested["deep"] != "[encrypted]" {
		t.Errorf("nested encrypted should be masked")
	}
	if nested["ok"] != "fine" {
		t.Errorf("nested plain should be unchanged")
	}
}
